### PR TITLE
Add demo for ARIA-256-CBC

### LINKED ISSUE #14104: Add an EVP non AEAD cipher demo
Suggestion: ARIA-256-CBC but not compulsory.

It should be the current best practice (fetch), have proper error handling and resource allocation/deallocation and be in the style of a known answer test.

### PR DESCRIPTION
**Pull request by faramir-dev**
Fixes #14104

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: demos/cipher/ariacbc.c [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    /* Fetch the cipher implementation */
+    if ((cipher = EVP_CIPHER_fetch(libctx, "ARIA-256-CBC", propq)) == NULL)
+        goto err;
+
+    /*
+     * Initialise an encrypt operation with the cipher/mode, key, IV and
+     * IV length parameter.
+     * For demonstration purposes the IV is being set here. In a compliant
+     * application the IV would be generated internally so the iv passed in
+     * would be NULL.
```
Comment (mattcaswell) [inline demos/cipher/ariacbc.c:84]: This is true for GCM mode. I don't think this is true generally (the IV is not generated automatically for most implementations)

File: demos/cipher/ariacbc.c [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    if ((ctx = EVP_CIPHER_CTX_new()) == NULL)
+        goto err;
+
+    /* Fetch the cipher implementation */
+    if ((cipher = EVP_CIPHER_fetch(libctx, "ARIA-256-CBC", propq)) == NULL)
+        goto err;
+
+    /*
+     * Initialise an encrypt operation with the cipher/mode, key, IV and
+     * IV length parameter.
```
Comment (mattcaswell) [inline demos/cipher/ariacbc.c:81]: You are not setting the IV length parameter so this part of the comment can be removed

File: demos/cipher/ariacbc.c [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+     * For demonstration purposes the IV is being set here. In a compliant
+     * application the IV would be generated internally so the iv passed in
+     * would be NULL. 
+     */
+    if (!(ret = EVP_EncryptInit_ex2(ctx, cipher, cbc_key, cbc_iv, params)))
```
Comment (mattcaswell) [inline demos/cipher/ariacbc.c:86]: You are not using any custom params so this can just be `NULL`

Comment (mattcaswell) [reply demos/cipher/ariacbc.c:86]: It should not be necessary to assign `ret` here

File: demos/cipher/ariacbc.c [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    if (!EVP_EncryptUpdate(ctx, outbuf, &outlen, cbc_pt, sizeof(cbc_pt)))
+        goto err;
+
+    /* Finalise: there can be some additional output from padding */
+    if (!(ret = EVP_EncryptFinal_ex(ctx, outbuf + outlen, &tmplen)))
+        goto err;
+    outlen += tmplen;
+
+    /* Output encrypted block */
+    printf("Ciphertext (outlen:%d; tmplen:%d):\n", outlen, tmplen);
```
Comment (mattcaswell) [inline demos/cipher/ariacbc.c:99]: Is it necessary to print tmplen here? Not sure what it adds

File: demos/cipher/ariacbc.c [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    if (!(ret = EVP_EncryptInit_ex2(ctx, cipher, cbc_key, cbc_iv, params)))
+        goto err;
+
+    /* Encrypt plaintext */
+    if (!EVP_EncryptUpdate(ctx, outbuf, &outlen, cbc_pt, sizeof(cbc_pt)))
+        goto err;
+
+    /* Finalise: there can be some additional output from padding */
+    if (!(ret = EVP_EncryptFinal_ex(ctx, outbuf + outlen, &tmplen)))
```
Comment (mattcaswell) [inline demos/cipher/ariacbc.c:94]: Should not need to assign `ret` here

File: demos/cipher/ariacbc.c [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    if ((ctx = EVP_CIPHER_CTX_new()) == NULL)
+        goto err;
+
+    /* Fetch the cipher implementation */
+    if ((cipher = EVP_CIPHER_fetch(libctx, "ARIA-256-CBC", propq)) == NULL)
+        goto err;
+
+    /*
+     * Initialise an encrypt operation with the cipher/mode, key, IV and
+     * IV length parameter.
```
Comment (mattcaswell) [inline demos/cipher/ariacbc.c:138]: Drop IV length

File: demos/cipher/ariacbc.c [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+    /* Finalise: there can be some additional output from padding */
+    if (!(ret = EVP_EncryptFinal_ex(ctx, outbuf + outlen, &tmplen)))
+        goto err;
+    outlen += tmplen;
+
+    /* Output encrypted block */
+    printf("Ciphertext (outlen:%d; tmplen:%d):\n", outlen, tmplen);
+    BIO_dump_fp(stdout, outbuf, outlen);
+
```
Comment (mattcaswell) [inline demos/cipher/ariacbc.c:101]: Should we compare against the expected ciphertext?

File: demos/cipher/ariacbc.c [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    if (!EVP_DecryptUpdate(ctx, outbuf, &outlen, cbc_ct, sizeof(cbc_ct)))
+        goto err;
+
+    /* Finalise: there can be some additional output from padding */
+    if (!EVP_DecryptFinal_ex(ctx, outbuf + outlen, &tmplen))
+        goto err;
+    outlen += tmplen;
+
+    /* Output decrypted block */
+    printf("Plaintext (outlen:%d; tmplen:%d):\n", outlen, tmplen);
```
Comment (mattcaswell) [inline demos/cipher/ariacbc.c:153]: Why print `tmplen`?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: demos/cipher/ariacbc.c (hunk 3 above)
Comment (faramir-dev) [reply demos/cipher/ariacbc.c:86]: `ret` is checked before return from the function, error message is printed if it is 0.

Comment (paulidale) [reply demos/cipher/ariacbc.c:81]: I think @mattcaswell just meant pass NULL here not params.

**Review by t8m — APPROVED** [review]

File: demos/cipher/ariacbc.c [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    /* Finalise: there can be some additional output from padding */
+    if (!EVP_EncryptFinal_ex(ctx, outbuf + outlen, &tmplen))
+        goto err;
+    outlen += tmplen;
+
+    /* Output encrypted block */
+    printf("Ciphertext (outlen:%d):\n", outlen);
+    BIO_dump_fp(stdout, outbuf, outlen);
+
+    if (sizeof(cbc_ct) == outlen && !memcmp(outbuf, cbc_ct, outlen))
```
Comment (mattcaswell) [inline demos/cipher/ariacbc.c:96]: Possibly should use `CRYPTO_memcmp` here? Its probably not strictly necessary in this instance but we should probably encourage users to use it when comparing crypto values

File: demos/cipher/ariacbc.c [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    /* Finalise: there can be some additional output from padding */
+    if (!EVP_DecryptFinal_ex(ctx, outbuf + outlen, &tmplen))
+        goto err;
+    outlen += tmplen;
+
+    /* Output decrypted block */
+    printf("Plaintext (outlen:%d):\n", outlen);
+    BIO_dump_fp(stdout, outbuf, outlen);
+
+    if (sizeof(cbc_pt) == outlen && !memcmp(outbuf, cbc_pt, outlen))
```
Comment (mattcaswell) [inline demos/cipher/ariacbc.c:152]: `CRYPTO_memcmp`?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks.

**Pull request closed without merging**